KAFKA-13202: KIP-768: Extend SASL/OAUTHBEARER with Support for OIDC

kirktrue opened the pull request:
This task is to provide a concrete implementation of the interfaces defined in KIP-255 to allow Kafka to connect to an OAuth/OIDC identity provider for authentication and token retrieval. While KIP-255 provides an unsecured JWT example for development, this will fill in the gap and provide a production-grade implementation.

The OAuth/OIDC work will allow out-of-the-box configuration by any Apache Kafka users to connect to an external identity provider service (e.g. Okta, Auth0, Azure, etc.). The code will implement the standard OAuth clientcredentials grant type.

The proposed change is largely composed of a pair of AuthenticateCallbackHandler implementations: one to login on the client and one to validate on the broker.

See the following for more detail:

* [KIP-768](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=186877575)
* [KAFKA-13202](https://issues.apache.org/jira/browse/KAFKA-13202)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
KAFKA-13202: KIP-768: Extend SASL/OAUTHBEARER with Support for OIDC

This task is to provide a concrete implementation of the interfaces defined in KIP-255 to allow Kafka to connect to an OAuth/OIDC identity provider for authentication and token retrieval. While KIP-255 provides an unsecured JWT example for development, this will fill in the gap and provide a production-grade implementation.

The OAuth/OIDC work will allow out-of-the-box configuration by any Apache Kafka users to connect to an external identity provider service (e.g. Okta, Auth0, Azure, etc.). The code will implement the standard OAuth clientcredentials grant type.

The proposed change is largely composed of a pair of AuthenticateCallbackHandler implementations: one to login on the client and one to validate on the broker.

See KIP-768: Extend SASL/OAUTHBEARER with Support for OIDC for more detail.

Commit:
Implement SSL socket factory

Implements the ability to configure and use a custom SSL socket factory
for any users who need to do so. Exposes the common SSL configuration
from the SslConfigs class. These can be configured within the login
module's JAAS settings.

Commit:
Added file watcher service and implemented PEM VerificationKeyResolver logic

Commit:
Implement AccessTokenRetriever with HTTP, file, and static string implementations

Commit:
OAuth/OIDC work clean up based on manual testing and verification

Commit:
Implement OAuth/OIDC standalone test harness

The OAuthCompatibilityTest configures the OAuth/OIDC layer (using system
properties), connects to the OAuth/OIDC provider to retrieve then token,
validates the JWT from the client perspective, then validates the JWT
from the broker perspective.

This test requires that an external OAuth/OIDC provider can be reached
from the system on which it is executed. It can be used for basic
verification of configuration and troubleshooting.

Commit:
Allowing the JWT parser/validator to accept any non-none alg header

Commit:
Removed watch service and fixed PEM and JWKS verification key resolvers

Commit:
Introduce factory-style creation for access token retrievers, access token validators, and verification key resolvers

Commit:
Changes based on Rajini's feedback on KIP

1. Configuration moved to top-level
2. Extension -> extension
3. JWKS on-demand "kid" lookups are non-blocking
4. Streamlined configuration by allowing jwksEndpointUri and
   tokenEndpointUri to specify file://-based URIs
5. Revised OAuthCompatibilityTest to use standard command line args

Commit:
Feedback from Ismael on KIP

Changed configuration names to match Kafka more closely
Introduced unretryable error handling for access token retrieval

Commit:
Fixed typo in KafkaConfig.scala

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerLoginCallbackHandler.java:
```diff
@@ -0,0 +1,188 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OAuthBearerLoginCallbackHandler implements AuthenticateCallbackHandler {
```
Could we add a high level description of the class?

kirktrue replied:
Yes, I have added appropriate JavaDoc describing this callback handler and contrasting it with the other.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerLoginCallbackHandler.java:
```diff
@@ -0,0 +1,188 @@
+        }
+    }
+
+    void handle(OAuthBearerTokenCallback callback) throws IOException {
```
Could this method and the next one be private?

kirktrue replied:
I believe these two methods were package visible because there were some tests that called them directly. Apparently those tests have been refactored. I'll change them to `private`.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/ConfigurationUtils.java:
```diff
@@ -0,0 +1,277 @@
+        this(configs, null);
+    }
+
+    public ConfigurationUtils(Map<String, ?> configs, String saslMechanism) {
```
saslMechanism seems to be used for the listener prefix. 
(1) Could we name it as listener name?
(2) The listener name is only relevant for the broker. Could we avoid passing in that for the client?
(3) Listener name is not always the sasl mechanism name. Should we pass in the listener name?

kirktrue replied:
Per the `AuthenticateCallbackHandler` interface, the `configure` method parameters do not include the listener name.

Based on my research, it seemed like the upstream code only passed in the configuration that's either listener-based or top-level.

The SASL mechanism is passed in for the case of the broker side. The later configuration lookups will look for the optional SASL mechanism prefix and use it if the configuration is there, or fall back to the top-level configuration.

kirktrue replied:
Or should I change the `AuthenticateCallbackHandler` interface's `configure` method to take the listener name too?

junrao replied:
It seems that the configs passed to the following method already has the listener prefix removed.

` void configure(Map<String, ?> configs, String saslMechanism, List<AppConfigurationEntry> jaasConfigEntries);
`

The logic does that is in ChannelBuilders.create(), where we have the following code.

`        Map<String, Object> configs = channelBuilderConfigs(config, listenerName);
`

So, the code in the PR is correct.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/ConfigurationUtils.java:
```diff
@@ -0,0 +1,277 @@
+        }
+
+        ConfigDef sslConfigDef = new ConfigDef();
+        sslConfigDef.withClientSslSupport();
```
Hmm, these are ssl properties for connections between Kafka client and Kafka broker. Here, we need the ssl properties for connecting to the oauth provider. They may require a different set of configurations.

kirktrue replied:
The client SSL configuration for OAuth comes from the JAAS options, not the configuration. So the user can specify SSL options like this:

```
sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required \
    clientId="foo" \
    clientSecret="bar" \
    scope="baz" \
    ssl.protocol="SSL"  \
    ssl.keystore.type="...." ;
```

This allows the HTTP client connection to use the same set of SSL configuration when connecting to the OAuth provider.

kirktrue replied:
That said, you're right, the code was accidentally referring to the top-level configuration `Map`, not the JAAS options `Map`. That has been fixed.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/ConfigurationUtils.java:
```diff
@@ -0,0 +1,277 @@
+        Map<String, ?> sslClientConfig = getSslClientConfig(uriConfigName);
+
+        if (sslClientConfig == null) {
+            log.warn("Requesting SSL client socket factory but SSL configs were null");
```
Do we need this warning since we already log a warn in getSslClientConfig()?

kirktrue replied:
I refactored the code a bit, so this has been removed.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/ConfigurationUtils.java:
```diff
@@ -0,0 +1,277 @@
+        URI uri;
+
+        try {
+            uri = new URI(value.trim());
```
value is already trimmed in validateString(). So, no need to trim it again.

kirktrue replied:
Extraneous `trim` removed.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/AccessTokenRetrieverFactory.java:
```diff
@@ -0,0 +1,78 @@
+                tokenEndpointUri.toString(),
+                cu.validateLong(SASL_LOGIN_RETRY_BACKOFF_MS),
+                cu.validateLong(SASL_LOGIN_RETRY_BACKOFF_MAX_MS),
+                cu.validateInteger(SASL_LOGIN_CONNECT_TIMEOUT_MS, false, null),
```
This could just use the constructor without the min value. Ditto below.

kirktrue replied:
Changed to use the two-parameter version of `validateInteger`.

YiDing-Duke commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java:
```diff
@@ -0,0 +1,325 @@
+
+        try (InputStream is = con.getInputStream()) {
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
+            log.debug("handleOutput - preparing to read request body to {}", con.getURL());
```
nit: should we say "response body" rather than "request body"?

kirktrue replied:
Yes 😱  Good catch!

YiDing-Duke commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java:
```diff
@@ -0,0 +1,325 @@
+
+            if (scope != null && !scope.trim().isEmpty()) {
+                scope = scope.trim();
+                String encodedScope = URLEncoder.encode(scope, StandardCharsets.UTF_8.name());
```
dummy question: does only the authorization header need to be Base64 encoded, the request body doesn't?

kirktrue replied:
The request body (parameters) are _URL_-encoded, but the `Authorization` header value after `Basic` needs to be _base 64_-encoded. That's per HTTP, I believe, not OAuth.

YiDing-Duke commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/Retry.java:
```diff
@@ -0,0 +1,104 @@
+                long diff = end - time.milliseconds();
+                waitMs = Math.min(waitMs, diff);
+
+                if (waitMs <= 0)
```
When it comes to retry timeout, should we log this error so that we can figure out a non retry-able error pattern to add to the list in the future?

kirktrue replied:
Added logging for error.

YiDing-Duke commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerLoginCallbackHandler.java:
```diff
@@ -0,0 +1,188 @@
+        configure(accessTokenRetriever, accessTokenValidator);
+    }
+
+    public void configure(AccessTokenRetriever accessTokenRetriever, AccessTokenValidator accessTokenValidator) {
```
nit: should we name this func as init()?

kirktrue replied:
Sure! I changed it to `init`.

YiDing-Duke commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerLoginCallbackHandler.java:
```diff
@@ -0,0 +1,188 @@
+        }
+    }
+
+    void handle(OAuthBearerTokenCallback callback) throws IOException {
```
nit: Should we name these two handle functions separetely? like handleCallback() and handleExtensionsCallback()?

kirktrue replied:
Sure! I changed it to be more specific.

YiDing-Duke commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,112 @@
+        configure(verificationKeyResolver, accessTokenValidator);
+    }
+
+    public void configure(CloseableVerificationKeyResolver verificationKeyResolver, AccessTokenValidator accessTokenValidator) {
```
ditto: should we use init() function name?

kirktrue replied:
Sure! I changed it to `init` here too.

YiDing-Duke commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,112 @@
+            throw new KafkaException("The OAuth validator configuration encountered an error when initializing the VerificationKeyResolver", e);
+        }
+
+        isConfigured = true;
```
nit: should we move the flag setting to the end of main configure() function?

kirktrue replied:
Unit tests will call `init` directly, which is OK. Changed the flag to `isInitialized` so it fits in better now.

YiDing-Duke commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,112 @@
+                handle((OAuthBearerValidatorCallback) callback);
+            } else if (callback instanceof OAuthBearerExtensionsValidatorCallback) {
+                OAuthBearerExtensionsValidatorCallback extensionsCallback = (OAuthBearerExtensionsValidatorCallback) callback;
+                extensionsCallback.inputExtensions().map().forEach((extensionName, v) -> extensionsCallback.valid(extensionName));
```
For now, the server side extension validation is a no-op, we just make it as validated? The real function will be our ce-kafka server side authentication?

kirktrue replied:
Correct.

Commit:
Implemented changes per Yi

Commit:
Merge pull request #12 from kirktrue/yi-ding-pr-review-changes

Implemented changes per Yi

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java:
```diff
@@ -0,0 +1,325 @@
+
+    private static String sanitizeString(String name, String value) throws IOException {
+        if (value == null)
+            throw new IOException(String.format("%s value must be non-null", name));
```
Hmm, this is not really an IOException. Perhaps it should be sth like IllegalArgument. Ditto below.

kirktrue replied:
Changed to `IllegalArgumentException` and updated relevant unit tests.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java:
```diff
@@ -0,0 +1,325 @@
+            if (responseBody == null || responseBody.isEmpty())
+                throw new IOException(String.format("The token endpoint response was unexpectedly empty despite response code %s from %s", responseCode, con.getURL()));
+
+            log.debug("handleOutput - responseCode: {}, response: {}", responseCode, responseBody);
```
Should we log this before throwing the IOException above?

kirktrue replied:
Moved the logging to before the exception is thrown.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java:
```diff
@@ -0,0 +1,325 @@
+        JsonNode accessTokenNode = rootNode.at("/access_token");
+
+        if (accessTokenNode == null)
+            throw new IOException("The token endpoint response did not contain an access_token value");
```
Should we include responseBody in the exception?

kirktrue replied:
I was/am a _little_ leery of including the response verbatim, only because the response could be huge, especially if the URL provided in the configuration is incorrect.

The code is now logging the entire response, but arbitrarily limiting the length to 1000 characters for the exception message.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java:
```diff
@@ -0,0 +1,325 @@
+            throw new IOException(String.format("%s value must be non-null", name));
+
+        if (value.isEmpty())
+            throw new IOException(String.format("%s value must be non-empty", name));
```
name -> value ?

kirktrue replied:
It is correct as written, though obviously confusing.

I changed the format string to:

```
The value for %s must be non-null
```

e.g.:

```
The value for the token endpoint request client ID parameter must be non-null
The value for the token endpoint response's access_token JSON attribute must not contain only whitespace
```

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java:
```diff
@@ -0,0 +1,325 @@
+        value = value.trim();
+
+        if (value.isEmpty())
+            throw new IOException(String.format("%s value must not contain only whitespace", name));
```
name -> value ?

kirktrue replied:
I adjusted the format string likewise, as per the above comment.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerLoginCallbackHandler.java:
```diff
@@ -0,0 +1,188 @@
+            throw new IllegalArgumentException(String.format("Must supply exactly 1 non-null JAAS mechanism configuration (size was %d)", jaasConfigEntries.size()));
+
+        moduleOptions = Collections.unmodifiableMap(jaasConfigEntries.get(0).getOptions());
+        AccessTokenRetriever accessTokenRetriever = AccessTokenRetrieverFactory.create(configs, saslMechanism, moduleOptions);
```
saslMechanism is used for listener prefix. Since listener is a broker side thing it seems that we don't need to pass this in the client? Ditto below.

kirktrue replied:
The login callback handler can be used from the broker side too.

The SASL mechanism is passed in for the case of the broker side. The later configuration lookups will look for the optional SASL mechanism prefix and use it if the configuration is there, or fall back to the top-level configuration.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/ValidatorAccessTokenValidator.java:
```diff
@@ -0,0 +1,209 @@
+
+/**
+ * ValidatorAccessTokenValidator is an implementation of {@link AccessTokenValidator} that is used
+ * by the broker to perform more extensive validation of the JWT access token that is received
```
It seems that ValidatorAccessTokenValidator is used for validation in the client too?

kirktrue replied:
Yes and no.

The `AccessTokenValidator` interface is used by both client and broker, but the `ValidatorAccessTokenValidator` implementation is specifically designed to only be used on the broker while the `LoginAccessTokenValidator` is used as the client implementation. 

The reason for this is that on the broker we have the ability to bundle utility libraries (like jose4j) that allow `ValidatorAccessTokenValidator` to perform more thorough validation of the JWT. We intentionally want to keep the client dependencies lightweight, so `LoginAccessTokenValidator` only performs basic parsing and sanity checking on the JWT.

I made an attempt to document this rationale in the `AccessTokenValidator` interface documentation. Let me know if there's something more specific I should put in there.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/ValidatorAccessTokenValidator.java:
```diff
@@ -0,0 +1,209 @@
+ *         <code>iat</code> claims
+ *     </li>
+ *     <li>
+ *         Signature matching validation against the <code>kid</code> and those provided b
```
provided b => provided by

kirktrue replied:
Thank ou for noticing that 😄 . Fixed.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/ValidatorAccessTokenValidator.java:
```diff
@@ -0,0 +1,209 @@
+     *                                broker will match the value from JWT's <code>aud</code>
+     *                                claim to see if there is an exact match. If there is no
+     *                                match, the broker will reject the JWT and authentication
+     *                                will fail. May be <code>null</code> to not perform an
```
`May be <code>null</code> to not perform an` seems incomplete.

kirktrue replied:
Completed the thought so that it now reads:

```
May be <code>null</code> to not perform any check to verify the JWT's <code>aud</code> claim matches any fixed set of known/expected audiences.
```

Thanks for catching that.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/ValidatorAccessTokenValidator.java:
```diff
@@ -0,0 +1,209 @@
+     *                                and if this value is set, the broker will match it exactly
+     *                                against what is in the JWT's <code>iss</code> claim.
+     *                                If there is no match, the broker will reject the JWT and
+     *                                authentication will fail.
```
What happens when expectedIssuer is null?

kirktrue replied:
I generically copied and pasted from above so that it now reads:

```
May be <code>null</code> to not perform any check to verify the JWT's <code>iss</code> claim matches a specific issuer.
```

YiDing-Duke commented on clients/src/test/java/org/apache/kafka/common/security/oauthbearer/secured/AccessTokenValidatorFactoryTest.java:
```diff
@@ -0,0 +1,77 @@
+        AccessTokenValidator accessTokenValidator = AccessTokenValidatorFactory.create(configs);
+        handler.configure(accessTokenRetriever, accessTokenValidator);
+
+        // Basically asserting this doesn't throw an exception :(
```
dummy question: why we don't want to throw exception when close fails?

kirktrue replied:
The `AuthenticateCallbackHandler` interface's `close` method doesn't allow for exceptions. So in our implementation, we just catch any errors and log them.

YiDing-Duke commented on clients/src/test/java/org/apache/kafka/common/security/oauthbearer/secured/AccessTokenValidatorFactoryTest.java:
```diff
@@ -0,0 +1,77 @@
+        handler.close();
+    }
+
+    private OAuthBearerLoginCallbackHandler createHandler(AccessTokenRetriever accessTokenRetriever, Map<String, ?> configs) {
```
This function is never used?

kirktrue replied:
Correct. I will remove it in the next revision.

YiDing-Duke commented on clients/src/test/java/org/apache/kafka/common/security/oauthbearer/secured/AccessTokenValidatorFactoryTest.java:
```diff
@@ -0,0 +1,77 @@
+        handler.close();
+    }
+
+    private OAuthBearerLoginCallbackHandler createHandler(AccessTokenRetriever accessTokenRetriever, Map<String, ?> configs) {
```
This function is never used?

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,112 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OAuthBearerValidatorCallbackHandler implements AuthenticateCallbackHandler {
```
Could we add a comment to this class?

kirktrue replied:
Yes, I have added appropriate JavaDoc describing this callback handler and contrasting it with the other.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,112 @@
+    @Override
+    public void configure(Map<String, ?> configs, String saslMechanism, List<AppConfigurationEntry> jaasConfigEntries) {
+        CloseableVerificationKeyResolver verificationKeyResolver = VerificationKeyResolverFactory.create(configs, saslMechanism);
+        AccessTokenValidator accessTokenValidator = AccessTokenValidatorFactory.create(configs, saslMechanism, verificationKeyResolver);
```
saslMechanism is used as listener prefix. It seems that we need to pass in the listener name too?

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,112 @@
+            log.debug("handle - token: {}", token);
+            callback.token(token);
+        } catch (ValidateException e) {
+            e.printStackTrace();
```
In general, we shouldn't print to stdout and should use logging instead.

kirktrue replied:
Whoopsie! Removed.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/JwksFileVerificationKeyResolver.java:
```diff
@@ -0,0 +1,79 @@
+
+/**
+ * <code>JwksFileVerificationKeyResolver</code> is a {@link VerificationKeyResolver} implementation
+ * that will load the JWKS from the given file system directory.
```
Could we describe the file format in code and the KIP?

kirktrue replied:
Done. Take a look and let me know if there's more that needs to be said.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/PemDirectoryVerificationKeyResolver.java:
```diff
@@ -0,0 +1,134 @@
+ * that will load the PEM files from the given file system directory.
+ *
+ * The instance is configured with the directory name that contains one or more
+ * <a href=\"https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail\">public key files</a>
```
Is https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail the right URL?

kirktrue replied:
I believe so 🤷 

From that Wikipedia page:

> Privacy-Enhanced Mail (PEM) is a de facto file format for storing and sending cryptographic keys, certificates, and other data.

and...

> PEM data is commonly stored in files with a ".pem" suffix, a ".cer" or ".crt" suffix (for certificates), or a ".key" suffix (for public or private keys).

kirktrue replied:
That said, I will strip this out as it's not needed for a first version.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/PemDirectoryVerificationKeyResolver.java:
```diff
@@ -0,0 +1,134 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * <code>PemVerificationKeyResolver</code> is a {@link VerificationKeyResolver} implementation
```
So, PemVerificationKeyResolver is only used for testing and not for production use?

kirktrue replied:
It was going to be a format that we support, but it doesn't need to be in the first version.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,239 @@
+
+            try {
+                refreshLock.writeLock().lock();
+                this.jsonWebKeys = localJWKs;
```
Hmm, I am a bit confused. What is jsonWebKeys intended for? Why is it set during init() but not after refresh()?

kirktrue replied:
Whoops! Refactoring gone wrong. It was meant to retrieve the JWKS outside of the lock (`super.getJsonWebKeys()`) and then assign it inside the write lock.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,239 @@
+            log.info("JWKS validation key refresh of {} complete", getLocation());
+        } catch (JoseException | IOException e) {
+            // Let's wait a random, but short amount of time before trying again.
+            long waitMs = ThreadLocalRandom.current().nextLong(1000, 10000);
```
Should we backoff exponentially up to the next refresh interval?

kirktrue replied:
Yes, that seems prudent. I have to think about this a bit more to make sure that the on-demand thread and the normally-scheduled background thread don't step on each other's toes.

kirktrue replied:
Implementation and tests for the backoff/retry now included.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,239 @@
+
+                // If there's no entry in the missing key ID cache for the incoming key ID,
+                // or it has expired, schedule a refresh.
+                Long lastCheckTime = missingKeyIds.get(keyId);
```
It seems that this is the next check time?

kirktrue replied:
Yes. I renamed the variable.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/Retry.java:
```diff
@@ -0,0 +1,106 @@
+ * @param <R> Result type
+ */
+
+public class Retry<R> {
```
This and Retryable are not oauth specific. Perhaps they can be put in common.utils?

kirktrue replied:
I'm a little hesitant to do so, only because I want to have freedom to refactor as needed.

junrao commented on clients/src/main/java/org/apache/kafka/common/config/SaslConfigs.java:
```diff
@@ -75,30 +78,111 @@
+            + " " + SASL_LOGIN_RETRY_BACKOFF_MS
+            + " setting and will double in wait length between attempts up to a maximum wait length specified by the"
+            + " " + SASL_LOGIN_RETRY_BACKOFF_MAX_MS
+            + " setting.";
```
Should we add OAUTHBEARER_NOTE?

kirktrue replied:
I have added the `OAUTHBEARER_NOTE` to the `EXPONENTIAL_BACKOFF_NOTE` so that it is included in the documentation for the `sasl.login.retry.backoff.ms` and `sasl.login.retry.backoff.max.ms` configuration options.

junrao commented on clients/src/main/java/org/apache/kafka/common/config/SaslConfigs.java:
```diff
@@ -75,30 +78,111 @@
+    // These are only specified here outside their normal groupings so that they can be
+    // forward referencing.
+    public static final String SASL_OAUTHBEARER_JWKS_ENDPOINT_URI = "sasl.oauthbearer.jwks.endpoint.uri";
+    public static final String SASL_OAUTHBEARER_JWKS_ENDPOINT_REFRESH_MS = "sasl.oauthbearer.jwks.endpoint.refresh.ms";
```
Should we keep this close to its default and doc?

kirktrue replied:
I have rearranged the ordering to be consistent with the existing structure. This meant that I had to reference the name in a string literal vs. using a reference to the name, but 🤷

junrao commented on build.gradle:
```diff
@@ -829,6 +829,7 @@ project(':core') {
     implementation libs.jacksonDataformatCsv
     implementation libs.jacksonJDK8Datatypes
     implementation libs.joptSimple
+    implementation libs.jose4j
```
Is this needed since core depends on clients, which already has this dependency?

kirktrue replied:
Clients intentionally *doesn't* have the jose4j dependency (to keep client dependencies to a minimum), so we have to explicitly include it here. 

Suggestions welcome on a better way to handle it.

junrao replied:
This is fine. It will be useful to check all other projects that depend on clients and see if the jose4j dependency needs to be added explicitly. For example, it's convenient to add it to shell and potentially connect:runtime.

junrao commented on build.gradle:
```diff
@@ -1604,6 +1608,7 @@ project(':tools') {
     implementation libs.slf4jApi
     implementation libs.log4j
 
+    implementation libs.jose4j
```
Is this needed since tools depends on clients, which already has this dependency?

kirktrue replied:
It is needed here as per the explanation for the need for the dependency in `:core`.

junrao reviewed (commented):
@kirktrue : Thanks for the PR. Made a pass of all non-testing files. A few comments below.

Commit:
PR review update #2

Commit:
Merge pull request #13 from kirktrue/kip-768-pr-seccond-pass

PR review update #2

Commit:
Added class-level documentation to callback handlers

Commit:
Removed PEM-based VerificationKeyResolver

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,181 @@
+ * <code>unsecuredLoginStringClaim_sub</code> is needed:
+ *
+ * <code>
+ * listener.name.<listener name>.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required \
```
Hmm, are OAuthBearerLoginModule and unsecuredLoginStringClaim_sub needed on the server side in the secure mode? The latter wasn't mentioned in the KIP.

kirktrue replied:
Ugh. Need to address. Thanks for reminding me 😄

kirktrue replied:
Outdated documentation removed.

Commit:
Updated SASL config ordering

Commit:
CHanged lastCheckTime to nextCheckTime, assigning jsonWebKeys in refreshInternal, and Changed isInited to isInitialized

Commit:
Added JWKS documentation to JwksFileVerificationKeyResolver

Commit:
Added JWKS documentation to RefreshingHttpsJwksVerificationKeyResolver too

kirktrue commented:
@junrao - I believe I've addressed all of the points listed except:

1. The listener/SASL mechanism prefix issues are really confusing to me because the code is currently behaving the way that I'm assuming it's supposed to. For example, if I have a listener named `foo` and an OAuth configuration named `bar`, `sasl.oauthbearer.bar=1` is picked up by my code unless there's a listener-specific override (e.g. `listener.name.foo.oauthbearer.sasl.oauthbearer.bar=2`). And if I specify a value for some other listener (e.g. `listener.name.someotherlistener.oauthbearer.sasl.oauthbearer.bar=3`, I don't see that value in the callback handler that is initialized for the `foo` listener.
2. I need to implement exponential backoff for the case of on-demand refresh of the JWKS.

Commit:
Modifications to Retry to support RefreshingHttpsJwks backoff

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,244 @@
+
+            try {
+                refreshLock.writeLock().lock();
+                this.jsonWebKeys = Collections.unmodifiableList(localJWKs);
```
`this.jsonWebKeys` could just be `jsonWebKeys`?

kirktrue replied:
Sure. I'll remove `this.`.

junrao commented on clients/src/test/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerTest.java:
```diff
@@ -0,0 +1,202 @@
+        return mockedCon;
+    }
+
+    protected File createTempPemDir() throws IOException {
```
This seems unused.

kirktrue replied:
It is, and it will be removed.

junrao commented on clients/src/test/resources/log4j.properties:
```diff
@@ -12,10 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
-log4j.rootLogger=OFF, stdout
+log4j.rootLogger=INFO, stdout
```
Are the changes in this file needed?

kirktrue replied:
No. This was me being careless 😱 

Sorry.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/LoginAccessTokenValidator.java:
```diff
@@ -0,0 +1,144 @@
+     * validation of the JWT.
+     */
+
+    public LoginAccessTokenValidator() {
```
This seems never used?

kirktrue replied:
Correct. It has been removed.

ijuma commented on build.gradle:
```diff
@@ -1200,6 +1201,7 @@ project(':clients') {
 
     compileOnly libs.jacksonDatabind // for SASL/OAUTHBEARER bearer token parsing
     compileOnly libs.jacksonJDK8Datatypes
+    compileOnly libs.jose4j
```
Do we want to add a comment like:

```text
// for SASL/OAUTHBEARER bearer token parsing
```

kirktrue replied:
Done!

junrao commented on clients/src/test/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerLoginCallbackHandlerTest.java:
```diff
@@ -0,0 +1,225 @@
+        OAuthBearerLoginCallbackHandler handler = new OAuthBearerLoginCallbackHandler();
+        Map<String, ?> configs = getSaslConfigs(SASL_OAUTHBEARER_TOKEN_ENDPOINT_URL, accessTokenFile.toURI().toString());
+        Map<String, Object> jaasConfigs = Collections.emptyMap();
+        configureHandler(handler, configs, jaasConfigs);
```
This test doesn't seem to verify anything. Ditto for the next test.

kirktrue replied:
Added proper assertions.

junrao commented on tools/src/main/java/org/apache/kafka/tools/OAuthCompatibilityTest.java:
```diff
@@ -0,0 +1,264 @@
+
+    private static Map<String, ?> getConfigs(Namespace namespace) {
+        Map<String, Object> c = new HashMap<>();
+        maybeAddLong(namespace, "connectTimeoutMs", c, SASL_LOGIN_CONNECT_TIMEOUT_MS);
```
This and the next one seem to be of type int.

kirktrue replied:
Wow! Nice catch. Changed to use `maybeAddInt` instead.

junrao reviewed (commented):
@kirktrue: Thanks for the updated PR. Made a pass of all file. A few more comments below.

Commit:
Feedback from Jun and Ismael on unused methods, incomplete tests, log4j changes, etc.

Commit:
Removed outdated documentation for OAuthBearerValidatorCallbackHandler

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,301 @@
+        log.debug("timeBeforeNextRefresh: {}, RETRY_BACKOFF_MS: {}", timeBeforeNextRefresh, RETRY_BACKOFF_MS);
+
+        // If the time left before the next scheduled refresh is less than the amount of time we
+        // have set aside for retries, log the fact and return. Don't worry, refreshInternal will
```
refreshInternal no longer exists.

kirktrue replied:
Will update comment. Good catch.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,301 @@
+            }
+
+            refreshFuture = executorService.scheduleAtFixedRate(this::refresh,
+                0,
```
Since we just retrieved from httpsJwks, should we set an initial delay?

kirktrue replied:
Yes. I will set the `initialDelay` parameter value to `refreshMs`.

junrao commented on build.gradle:
```diff
@@ -1208,6 +1210,8 @@ project(':clients') {
     testRuntimeOnly libs.jacksonDatabind
     testRuntimeOnly libs.jacksonJDK8Datatypes
+    testRuntimeOnly libs.jose4j
+    testImplementation libs.jose4j
```
Hmm, I thought that testImplementation covers testRuntimeOnly?

kirktrue replied:
It appears it does. Thanks!

junrao commented on build.gradle:
```diff
@@ -1604,6 +1608,7 @@ project(':tools') {
     implementation libs.slf4jApi
     implementation libs.log4j
 
+    implementation libs.jose4j                    // for SASL/OAUTHBEARER JWT validation
```
Should we add the same dependency to shell and connect:runtime?

kirktrue replied:
Added the dependency to `shell` and `connect.runtime` too.

junrao commented on clients/src/test/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwksTest.java:
```diff
@@ -0,0 +1,135 @@
+        return httpsJwks;
+    }
+
+    private JsonWebKey mockJsonWebKey(String keyId) {
```
This seems unused.

kirktrue replied:
It was, then wasn't, then was, and now isn't again. I will remove it. Thanks.

junrao reviewed (commented):
@kirktrue : Thanks for the updated PR. A few more minor comments.

Commit:
More changes to JWKS refresh tests and feedback from Jun

Commit:
Updated build.gradle dependencies for clients, shell, and connect:runtime

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,298 @@
+            List<JsonWebKey> localJWKs;
+
+            try {
+                // This will trigger a call th HttpsJwks.refresh() and that will block the current
```
th => the

kirktrue replied:
Revised comments so this typo should be gone now.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,298 @@
+            try {
+                // This will trigger a call th HttpsJwks.refresh() and that will block the current
+                // thread. It's OK to do this in init() but we avoid blocking elsewhere as we're
+                // run in a network thread.
```
run => running

kirktrue replied:
"run" sounds right to me, but I'll change it to "running" because I tend to forgot rules of grammar 😄

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,298 @@
+        // will still be called again within RETRY_BACKOFF_MS :)
+        //
+        // Note: timeBeforeNextRefresh is negative when we're in the midst of executing refresh
+        // in a scheduled fashion. ScheduledFuture.getDelay will reset *after* the method has
```
Hmm, if we're in the midst of executing refresh, it seems that should return immediately instead of refreshing?

kirktrue replied:
I'm not sure I fully understand the question, sorry. Can you rephrase?

`refresh` can get called in one of two ways:

1. Scheduled refresh
2. Expedited refresh

An expedited refresh is performed because the JWT included a key ID that we don't have in our cache. The idea I implemented was to skip execution of an expedited refresh if the amount of time before the next scheduled refresh was less than some set amount. Basically: if we're going to do a scheduled refresh in the next N seconds or so, let's just wait for that. Kind of a cheap batching idea.

Another option is to just perform the refresh without any of that checking, but have a flag (`isCurrentlyRefreshing`) that is checked at the start of a refresh and returns if there's already one executing. At least that's how I'm interpreting your comment.

junrao replied:
If timeBeforeNextRefresh <= 0, it seems that we will trigger a refresh. However, in that case, we're in the midst of executing refresh. So, another refresh seems unnecessary?

kirktrue replied:
Changed the logic to be simpler: if we're in the middle of a `refresh` in thread `A` and thread `B` starts to execute `refresh` (either scheduled or expedited), we simply let thread `B` return. This is implemented via a quick CAS at the start of `refresh` using an `AtomicBoolean`.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,298 @@
+
+    public boolean maybeScheduleRefreshForMissingKeyId(String keyId) {
+        if (keyId.length() > MISSING_KEY_ID_MAX_KEY_LENGTH) {
+            // Only grab the first N characters so that if the key ID is huge, we don't blow up
```
don't blow up => will blow up ?

kirktrue replied:
Changed to:

```
// Although there's no limit on the length of the key ID, they're generally
// "reasonably" short. If we have a very long key ID length, we're going to assume
// the JWT is malformed and we will not actually try to resolve the key.
// 
// In this case, let's prevent blowing out our memory in two ways:
// 
//     1. Don't try to resolve the key as the large ID will sit in our cache
//     2. Report the issue in the logs but include only the first N characters
```

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,298 @@
+ * @see ValidatorAccessTokenValidator
+ */
+
+public final class RefreshingHttpsJwks implements Initable, Closeable {
```
A meta question, currently, it seems that each processor creates its own sasl callbacks, each of which then creates its own RefreshingHttpsJwks. This means that each processor will make its own http requests and maintain its own key cache. This could add overhead. It seems that a better approach is to have all processors for the same listener share the http requests and the key cache. What do you think?

kirktrue replied:
Yes, that seems reasonable.

The trick is that the `AuthenticateCallbackHandler` doesn't get the listener name, so it's hard to know when to share and when to create a new instance.

I'll look into the implementation.

kirktrue replied:
I implemented the change you requested by adding the appropriate code to `OAuthBearerValidatorCallbackHandler` reuse shared instances.

junrao reviewed (commented):
@kirktrue : Thanks for the updated PR. A few more comments below.

Commit:
More changes to JWKS refresh comments

Commit:
Fixed some comments

Commit:
Implemented shared VerificationKeyResolver so that we only have a single instance to reduce threads and HTTP(S) lookups

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,283 @@
+     * to keep a single instance per key.
+     */
+
+    private static class VkrKey {
```
What does vkr stand for?

kirktrue replied:
Stands for `VerificationKeyResolver`, but I will spell it out as it's not obvious to everyone.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,283 @@
+
+        // Here's the logic which keeps our VKRs down to a single instance.
+        synchronized (VKR_CACHE) {
+            VkrKey key = new VkrKey(configs, saslMechanism, moduleOptions);
```
Do we need saslMechanism as part of the key? It seems that saslMechanism should always be oauthbearer. Perhaps we could assert that?

kirktrue replied:
We do assert that the SASL mechanism is `oauthbearer` in a previous step, so I'll just remove `saslMechanism` as part of the key.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,283 @@
+    /**
+     * Because a {@link CloseableVerificationKeyResolver} instance can spawn threads and issue
+     * HTTP(S) calls ({@link RefreshingHttpsJwksVerificationKeyResolver}), we only want to create
+     * a new instance for each particular set of configuration. Because each of set of configuration
```
each of set => each set

kirktrue replied:
Changed.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,328 @@
+    /**
+     * Our implementation avoids the blocking call within {@link HttpsJwks#refresh()} that is
+     * sometimes called internal to {@link HttpsJwks#getJsonWebKeys()}. We want to avoid any
+     * blocking I/O is this code is running in the authentication path on the Kafka network thread.
```
is this code => as this code

kirktrue replied:
Changed.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/RefreshingHttpsJwks.java:
```diff
@@ -0,0 +1,328 @@
+     * Internal method that will refresh the cache in a background thread.
+     *
+     * This method may be called as part of the <i>normally</i>-scheduled refresh call
+     * (via {@link #executorService}, but it may have
```
"but it may have" seems incomplete.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,283 @@
+ * like so:
+ *
+ * <code>
+ * listener.name.<listener name>.oauthbearer.sasl.server.callback.handler.class=org.apache.kafka.common.security.oauthbearer.secured.OAuthBearerValidatorCallbackHandler
```
The KIP uses listener.name.<listener name>.oauthbearer.sasl.login.callback.handler.class, could we fix that?

kirktrue replied:
Updated the KIP. Thanks.

junrao reviewed (commented):
@kirktrue : Thanks for the updated PR. A few more comments below.

Commit:
Updated comments and concurrent refresh prevention logic around JWKS verification

junrao commented on clients/src/main/java/org/apache/kafka/common/config/SaslConfigs.java:
```diff
@@ -75,30 +78,116 @@
+    public static final long DEFAULT_SASL_LOGIN_RETRY_BACKOFF_MAX_MS = 10000;
+    public static final String SASL_LOGIN_RETRY_BACKOFF_MAX_MS_DOC = "The (optional) value in milliseconds for the maximum wait between login attempts to the"
+            + " external authentication provider."
+            + EXPONENTIAL_BACKOFF_NOTE;
```
EXPONENTIAL_BACKOFF_NOTE refers to sasl.login.retry.backoff.max.ms, which will be confusing. Ditto for  SASL_LOGIN_RETRY_BACKOFF_MS_DOC.

kirktrue replied:
Renamed `EXPONENTIAL_BACKOFF_NOTE` to `LOGIN_EXPONENTIAL_BACKOFF_NOTE` with the existing message.

Added `JWKS_EXPONENTIAL_BACKOFF_NOTE` which reads:

```
JWKS retrieval uses an exponential backoff algorithm with an initial wait based on
the sasl.oauthbearer.jwks.endpoint.retry.backoff.ms setting and will double in wait
length between attempts up to a maximum wait length specified by the
sasl.oauthbearer.jwks.endpoint.retry.backoff.max.ms setting.
```

Updated `SASL_OAUTHBEARER_JWKS_ENDPOINT_RETRY_BACKOFF_MS` and `SASL_OAUTHBEARER_JWKS_ENDPOINT_RETRY_BACKOFF_MAX_MS` to reference `JWKS_EXPONENTIAL_BACKOFF_NOTE`.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/OAuthBearerValidatorCallbackHandler.java:
```diff
@@ -0,0 +1,278 @@
+     * appropriate number of times.
+     */
+
+    private static class RefCountingVerificationKeyResolve implements CloseableVerificationKeyResolver {
```
RefCountingVerificationKeyResolve => RefCountingVerificationKeyResolver?

kirktrue replied:
Will make your suggested change.

junrao reviewed (commented):
@kirktrue : Thanks for the updated PR. Just a couple of more minor comments.

Commit:
Config doc change and class typo rename

junrao reviewed (approved):
@kirktrue : Thanks for the latest PR. LGTM. Are the test failures related to this PR? Also, it seems that we added a couple of more new configs not covered in the KIP. It would be useful to update the KIP wiki and update the KIP thread about the changes.

YiDing-Duke reviewed (approved):
Thank you for all the update. LGTM, assuming all the test failures are not relate to the PR.

Commit:
Fixed KafkaConfigTest failures by adding necessary configuration options
